bundle/edu.gemini.pit/src/main/scala/edu/gemini/pit/ui/editor/SimbadLookup.scala

```diff
@@ -0,0 +1,100 @@
+package edu.gemini.pit.ui.editor
+
+import edu.gemini.model.p1.immutable.{ Target, SiderealTarget }
+import edu.gemini.pit.catalog.{ Result => SimbadResult, Simbad }
+import edu.gemini.pit.catalog.{ Success, NotFound, Error, Offline }
+import edu.gemini.shared.gui.GlassLabel
+import java.awt.Component
+import java.util.logging.Logger
+import javax.swing.JOptionPane
+import jsky.util.gui.DialogUtil
+import scala.swing.Swing
+import scalaz._, Scalaz._, scalaz.concurrent.Task
+
+/** Synchronous wrapper for the UI aspects of a Simbad lookup. */
+final class SimbadLookup(editor: TargetEditor) {
+  private[this] val Log = Logger.getLogger(getClass.getName)
+  private[this] var glass = Option.empty[GlassLabel]
+
+  /** Add an `ensuring` method to Task so it works like IO. */
+  private implicit class MoreTaskOps[A](ta: Task[A]) {
+    def ensuring(always: Task[Unit]): Task[A] =
+      ta.attempt.flatMap {
+        case -\/(t) => always *> Task.fail(t)
+        case \/-(r) => always.as(r)
+      }
+  }
+
+  /** Perform a search with the given query, prompting for disambiguation if
+    * needed, replacing the TargetEditor on success.
+    * @param query the horizons query
+    */
+  def lookup(query: String): Unit = {
+    val action = search(query).ensuring(hide)
+    action.unsafePerformAsync {
+      case -\/(t)  => Swing.onEDT(DialogUtil.error(editor.peer, t))
+      case \/-(()) => ()
+    }
+  }
+
+  private def search(query: String): Task[Unit] =
+    show("Searching...") *> Task.async[SimbadResult] { k =>
+      Simbad.find(query)(r => k(r.right))
+    } >>= {
+
+      // Failure cases
+      case Offline     => fail("SIMBAD is offline or unreachable.")
+      case NotFound(d) => fail("No results found.")
+      case Error(t)    => Task.fail(t) // roll into top-level handler
+
+      // Success case will always be Nil on the right for a VOTableCatalog.
+      // This is a bad design but ¯\_(ツ)_/¯
```

Comment:
Why's that?

Reply:
The types need to be more precise. It's basically `Ior[Nel[A], Nel[B]]` so there are three cases to consider but they're all folded into one here, with no non-empty constraint. But it works and improving it isn't in the scope of this issue.